lib/terraform/runner.rb

```diff
@@ -0,0 +1,293 @@
+
+        RestClient::Resource.new(
+          server_url,
+          :headers    => {:authorization => "Bearer #{server_token}"},
+          :verify_ssl => verify_ssl
+        )
+      end
+
+      def stack_tenant_id
+        # TODO: fix hardcoded tenant_id
+        'c158d710-d91c-11ed-9fee-d93323035b4e'
+      end
+
+      # Create TerraformRunner Stack Job
+      def create_stack_job(
+        template_path,
+        input_vars: [],
+        tags: nil,
+        credentials: [],
+        env_vars: {},
+        name: "stack-#{rand(36**8).to_s(36)}"
+      )
+        _log.info("start stack_job for template: #{template_path}")
+        tenant_id = stack_tenant_id
+
+        # Temp Zip File
+        # zip_file_path = Tempfile.new(%w/tmp .zip/).path
+        zip_file_path = Tempfile.new(%w[tmp .zip]).path
+        create_zip_file_from_directory(zip_file_path, template_path)
+        zip_file_hash = Base64.encode64(File.binread(zip_file_path))
+
+        # TODO: use tags,env_vars
+        payload = JSON.generate(
+          {
+            :cloud_providers => credentials,
+            :name            => name,
+            :tenantId        => tenant_id,
+            :templateZipFile => zip_file_hash,
+            :parameters      => convert_to_cam_parameters(input_vars)
+          }
+        )
+        # _log.debug("Payload:>\n, #{payload}")
+        http_response = terraform_runner_client['api/stack/create'].post(
+          payload, :content_type => 'application/json'
+        )
+        _log.debug("==== http_response.body: \n #{http_response.body}")
+        _log.info("stack_job for template: #{template_path} running ...")
+        Terraform::Runner::Response.parsed_response(http_response)
+      ensure
+        # cleanup temp zip file
+        FileUtils.rm_rf(zip_file_path) if zip_file_path
+        _log.debug("Deleted #{zip_file_path}")
+      end
+
+      # Retrieve TerraformRunner Stack Job details
+      def retrieve_stack_job(stack_id)
+        payload = JSON.generate(
+          {
+            :stack_id => stack_id
+          }
+        )
+        http_response = terraform_runner_client['api/stack/retrieve'].post(
+          payload, :content_type => 'application/json'
+        )
+        _log.info("==== Retrieve Stack Response: \n #{http_response.body}")
+        Terraform::Runner::Response.parsed_response(http_response)
+      end
+
+      # Cancel/Stop running TerraformRunner Stack Job
+      def cancel_stack_job(stack_id)
+        payload = JSON.generate(
+          {
+            :stack_id => stack_id
+          }
+        )
+        http_response = terraform_runner_client['api/stack/cancel'].post(
+          payload, :content_type => 'application/json'
+        )
+        _log.info("==== Cancel Stack Response: \n #{http_response.body}")
+        Terraform::Runner::Response.parsed_response(http_response)
+      end
+
+      # Wait for TerraformRunner Stack Job to complete
+      def wait_until_completes(stack_id)
```

Comment:
@putmanoj will this method actually be used?  The preferred way of doing this is by using the MiqQueue to check the status and if it isn't completed to requeue itself (aka the Job state machine).

This has the potential to timeout the queue worker if it takes too long.

Reply:
I agree.  @agrare and I chatted about this and we think it makes sense to only have `run_async` and `run` can call `run_async`.  We think `run` (synchronously) was written for ansible because the CLI had that option and it would handle the waiting.  We don't really need this here.  If that's fine, I can remove this and make the changes so `run` is just `run_async`.

Reply:
let me know what you think @putmanoj

Reply:
There always will be CREATE action, for starting a Job in terraform-runner  and checking the Job status

Reply:
I've removed run being run synchronously and now it's an alias of run_async.  Let me know if I didn't do it right.